### DISCUSSION TITLE
If not using Maui, is there a WinUI 3 equivalent to SKGLView?

### DISCUSSION THREAD

**Discussion by ToolmakerSteve** [mono/SkiaSharp#3073] [Q&A] [answered]
For a project that only runs on Windows 11, is there a way to use SkiaSharp and OpenGL/Angle 3D directly in WinUI 3?

In SkiaSharp 3 Preview, I see an AngleSwapChainPanel. But that seems to be a low-level entity. Not many properties/methods. (I don't know much about WinUI 3 GUI programming.)

I'll probably use Maui and create a Window and an SKGLView, but I am wondering if there is some more direct way to program Windows 11 GUI. This is primarily a test to get a better sense of what dependencies get pulled in by relying on Maui.

**Comment by mattleibow** [comment, marked as answer]
There is a SKSwapChainPanel for WinUI I. The SkiaSharp.Views.WinUI nuget. No need for maui.

**Reply by mattleibow** [reply]
But maui gives you free mac for desktop and android and ios for mobile.

**Reply by ToolmakerSteve** [reply]
tl;dr Yes, we will use Maui. At least as a wrapper to host SkiaSharp.  
[Next UI topic will be another discussion.]

Details:

My current client was badly burned by dependency on Xamarin Forms.  
Previously abandoned a Xamarin Forms version, and went back to older custom UI code.  
Summary of that experience:
* Xamarin with native APIs on mobile platforms: Awesome. "C# everywhere" - game changing.
* Xamarin Forms: Problematic.

Due to this expensive misstep, and Microsoft's history of abandoning some technologies, Client is nervous about the long term future of Maui.  
[I personally would go with Maui Controls for UI; but it may be "too little, too late"; too many developers have moved on to other frameworks. I'm not going to recommend it to a business client who is hesitant.]

Originally this was .Net Framework on WPF; "Xamarin Native" on Android, iOS. A separate project per platform.  
My partner successfully converted their custom UI (which he originally wrote in C++ and GL shaders, years ago) to SkiaSharp.  
The existence of SkiaSharp is a lifesaver for us! Thank you so much for your dedication to that.

And we've converted to .Net 8.  
In .Net 8+, the underpinnings of multi-platform development looks solid.  
Gained confidence in long term viability, at least at that level.
